File: 053_MinimumCostToConnectSticks.py
"""
Topic: Heap / Greedy
Problem: Minimum Cost to Connect Sticks
Description: Combine smallest two repeatedly using min-heap.
Time: O(n log n)
Space: O(n)
"""

import heapq
from typing import List


def connectSticks(sticks: List[int]) -> int:
    if not sticks:
        return 0
    heapq.heapify(sticks)
    cost = 0
    while len(sticks) > 1:
        a = heapq.heappop(sticks)
        b = heapq.heappop(sticks)
        cost += a + b
        heapq.heappush(sticks, a + b)
    return cost


if __name__ == '__main__':
    print(connectSticks([2,4,3]))  # 14
    print(connectSticks([1,8,3,5]))  # 30

File: 053_MinimumCostToConnectSticks.cpp
// Topic: Heap / Greedy
// Minimum Cost to Connect Sticks: repeatedly connect two smallest sticks; cost is sum.
// Time Complexity: O(n log n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

long long connectSticks(vector<int>& sticks)
{
    priority_queue<long long, vector<long long>, greater<long long>> pq(sticks.begin(),
                                                                        sticks.end());
    long long cost = 0;
    while (pq.size() > 1)
    {
        long long a = pq.top();
        pq.pop();
        long long b = pq.top();
        pq.pop();
        cost += a + b;
        pq.push(a + b);
    }
    return cost;
}

int main()
{
    cout << connectSticks((vector<int>{2, 4, 3})) << "\n";     // 14
    cout << connectSticks((vector<int>{1, 8, 3, 5})) << "\n";  // 30
    return 0;
}
